### PR TITLE
Return an error code if someone uses a wrong filter

### PR DESCRIPTION
**Pull request by lboyarsky**
Ticket: [Return an error code if someone uses a wrong filter](https://app.asana.com/0/810933294009540/1104657015105420)

### CODE REVIEW

**Comment by lboyarsky** [comment]
Sorry I missed few controllers, which handle includes differently from others. It's now fixed. 
As for `?hello=world`, the ticket didn't ask to address such things unless I'm missing something.

**Comment by paulswartz** [comment]
`?hello=world` would be the same as `?filter[hello]=world`: for example, `vehicles?route=9` and `vehicles?filter[route]=9` do the same thing.

File: apps/api_web/lib/api_web/params.ex [hunk 1]
```diff
@@ -256,23 +256,60 @@ defmodule ApiWeb.Params do
...
+        end
+
+      _ ->
+        {:ok, %{}}
+    end
+  end
+
+  @spec validate_includes(map, [String.t()]) :: {:ok, [String.t()]} | {:error, atom}
+  def validate_includes(params, includes) do
+    case params["include"] do
```
Comment (paulswartz) [inline apps/api_web/lib/api_web/params.ex]: `Map.get(params, "include")` (no need to go through the access protocol)

File: apps/api_web/lib/api_web/params.ex [hunk 2]
```diff
@@ -256,23 +256,60 @@ defmodule ApiWeb.Params do
...
+      values when is_binary(values) ->
+        split =
+          values
+          |> String.split(",", trim: true)
+          |> Enum.map(&(&1 |> String.split(".") |> List.first()))
+
+        if split -- includes == [] do
+          {:ok, split}
+        else
+          {:error, :bad_include}
```
Comment (paulswartz) [inline apps/api_web/lib/api_web/params.ex]: Can you include the bad values as a part of the error? it'll be a more helpful message.

File: apps/api_web/test/api_web/controllers/stop_controller_test.exs [hunk 3]
```diff
@@ -206,22 +206,6 @@ defmodule ApiWeb.StopControllerTest do
       end
     end
 
-    test "can include a route we're filtering on", %{conn: conn} do
```
Comment (paulswartz) [inline apps/api_web/test/api_web/controllers/stop_controller_test.exs]: This shouldn't be removed: "/stops?filter[route]=Red&include=route` should work.

File: apps/api_web/lib/api_web/params.ex [hunk 4]
```diff
@@ -256,23 +256,60 @@ defmodule ApiWeb.Params do
...
     case params["filter"] do
-      filter when is_map(filter) -> Map.take(filter, keys)
-      _ -> %{}
+      filter when is_map(filter) ->
+        filtered = Map.take(filter, keys)
+
+        if map_size(filtered) == map_size(filter) do
+          {:ok, filtered}
+        else
+          {:error, :bad_filter}
```
Comment (paulswartz) [inline apps/api_web/lib/api_web/params.ex]: Can you include the bad filter values? It'll be a better error message.

File: apps/api_web/lib/api_web/params.ex (hunk 1 above)
Comment (lboyarsky) [reply apps/api_web/lib/api_web/params.ex]: Fixed

File: apps/api_web/lib/api_web/params.ex (hunk 2 above)
Comment (lboyarsky) [reply apps/api_web/lib/api_web/params.ex]: Done

File: apps/api_web/test/api_web/controllers/stop_controller_test.exs (hunk 3 above)
Comment (lboyarsky) [reply apps/api_web/test/api_web/controllers/stop_controller_test.exs]: Ok, it's just that swagger didn't had "route" as include options, so I though it was obsolete. Fixed.

File: apps/api_web/lib/api_web/params.ex (hunk 4 above)
Comment (lboyarsky) [reply apps/api_web/lib/api_web/params.ex]: Done

**Comment by lboyarsky** [comment]
Ready for the new review.

**Comment by lboyarsky** [comment]
Can you try again? Or I can try myself.

**Comment by lboyarsky** [comment]
I've deployed the fix to dev-green (this is what dev website uses, right?) and I don't see 400 spike any more. 

I see some 400s from requests like https://api-dev-green.us-east-1.elasticbeanstalk.com/stops?filter[date]=1987-03-31&filter[direction_id]=1&fields=whatever?include=parent_stations,facilities&rorute=Green-B, they return `no_service` error, but it's the same error you would receive on production with these query parameters.

**Comment by paulswartz** [comment]
The dev website uses `dev.api.mbtace.com`: you can deploy to `Dev` from Semaphore to try it out.

File: apps/api_web/test/api_web/controllers/stop_controller_test.exs (hunk 3 above)
Comment (paulswartz) [reply apps/api_web/test/api_web/controllers/stop_controller_test.exs]: Can you add that, either as a part of this or in a new ticket?

**Comment by lboyarsky** [comment]
@paulswartz I've checked and indeed they're using invalid (as far as I can tell) filters: 
- (~50% of errors) vehicle controller doesn't support filtering by date:

> 2019-02-27T15:13:53.783 [info]  node=api-i-04b9f794a48d5f21e@1bec4250bf97 request_id=2m3k2lf2jonk19art80005a2 api_key=418e5fda582e4217a667bf8f0d50476d state=set duration=0.775 status=400 params={"api_key":"418e5fda582e4217a667bf8f0d50476d","filter":{"date":"2019-02-27","route":"Shuttle002,Shuttle005"},"include":"route,stop"} path=/vehicles method=GET format=json-api controller=ApiWeb.VehicleController action=index api_version=2017-11-28 api_key=418e5fda582e4217a667bf8f0d50476d request_id=2m3k2lf2jonk19art80005a2

- (~40% of errors) shape controller doesn't support filtering by date: 
> 2019-02-27T15:13:54.096 [info]  node=api-i-04b9f794a48d5f21e@1bec4250bf97 request_id=2m3k2lfl93a1mrjka40005e2 api_key=418e5fda582e4217a667bf8f0d50476d state=set duration=0.789 status=400 params={"api_key":"418e5fda582e4217a667bf8f0d50476d","filter":{"date":"2019-02-27","route":"Shuttle002,Shuttle005"},"include":"stops"} path=/shapes method=GET format=json-api controller=ApiWeb.ShapeController action=index api_version=2017-11-28 api_key=418e5fda582e4217a667bf8f0d50476d request_id=2m3k2lfl93a1mrjka40005e2

What should I do about it?

**Comment by paulswartz** [comment]
Can you make a PR for them?

**Comment by lboyarsky** [comment]
https://github.com/mbta/shuttle-tracker/pull/5

File: apps/api_web/test/api_web/controllers/stop_controller_test.exs (hunk 3 above)
Comment (lboyarsky) [reply apps/api_web/test/api_web/controllers/stop_controller_test.exs]: I've added it to the list of includes in previous commit -- https://github.com/mbta/api/pull/11/commits/0bce0c32c06bef24ff50f09d7e68fe2610ef366f#diff-cee2a701bc0c45852758feec0c9de777L9 -- and it seems to be working without requiring additional code changes: 

```bash
levs-mbp:~ lev$ curl 'http://localhost:4000/stops?route=Mattapan&include=route' | jsonpp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4426  100  4426    0     0  85970      0 --:--:-- --:--:-- --:--:-- 86784
{
  "data": [
    {
      "attributes": {
        "address": "Dorchester Ave and Ashmont St, Boston, MA 02124",
        "description": null,
        "latitude": 42.284652,
        "location_type": 1,
        "longitude": -71.064489,
        "name": "Ashmont",
        "platform_code": null,
        "platform_name": null,
        "wheelchair_boarding": 1
      },
      "id": "place-asmnl",
      "links": {
        "self": "/stops/place-asmnl"
      },
      "relationships": {
        "child_stops": {},
        "facilities": {
          "links": {
            "related": "/facilities/?filter[stop]=place-asmnl"
          }
        },
        "parent_station": {
          "data": null
        },
        "route": {
          "data": {
            "id": "Mattapan",
            "type": "route"
          }
        }
      },
      "type": "stop"
    }
[...]
```

Comment (paulswartz) [reply apps/api_web/test/api_web/controllers/stop_controller_test.exs]: Sorry, I meant to add it to Swagger.

Comment (lboyarsky) [reply apps/api_web/test/api_web/controllers/stop_controller_test.exs]: Adding to Swagger is what I did in the commit linked above. 

image

File: apps/api_web/test/api_web/params_test.exs [hunk 5]
```diff
@@ -58,25 +58,34 @@ defmodule ApiWeb.ParamsTest do
...
-      assert Params.filter_params(params, ["stop"]) == %{"stop" => "4,5,6"}
+    test "returns error for invalid includes" do
+      assert Params.validate_includes(%{"include" => "stops,routes"}, ~w(stops trips)) ==
+               {:error, :bad_include, ~w(routes)}
     end
 
-    test "invalid filter values are removed" do
-      assert Params.filter_params(%{"filter" => "string"}, ["stop"]) == %{}
+    test "supports dot notation" do
+      assert Params.validate_includes(%{"include" => "stops.id"}, ~w(stops routes trips)) ==
```
Comment (paulswartz) [inline apps/api_web/test/api_web/params_test.exs]: shouldn't this be `{:error, :bad_include, ~w(id)}` ?

Comment (paulswartz) [reply apps/api_web/test/api_web/params_test.exs]: Maybe it's more trouble than it's worth, though.

Comment (lboyarsky) [reply apps/api_web/test/api_web/params_test.exs]: Yes, I thought that validation of the top level part (that is, everything before the dot) should generally be enough in terms of maintaining value vs complexity balance. Let me know if you want me to look into the implementation of additional validations.

Comment (paulswartz) [reply apps/api_web/test/api_web/params_test.exs]: No, I think this should be good for now.

**Pull request merged**